### ISSUE TITLE
Propagation failure for highly eccentric hyperbolic orbits with mean_motion

### ISSUE THREAD

**Issue by Juanlu001** [poliastro/poliastro#347]
🐞 **Problem**

The last cell of the [Florence notebook](http://docs.poliastro.space/en/latest/examples/Catch%20that%20asteroid!.html) fails with these warnings and errors:

```
/home/juanlu/.miniconda36/envs/poliastro36/lib/python3.6/site-packages/astropy/units/quantity.py:639: RuntimeWarning:

overflow encountered in cosh

/home/juanlu/.miniconda36/envs/poliastro36/lib/python3.6/site-packages/astropy/units/quantity.py:639: RuntimeWarning:

overflow encountered in sinh

/home/juanlu/.miniconda36/envs/poliastro36/lib/python3.6/site-packages/astropy/units/quantity.py:639: RuntimeWarning:

invalid value encountered in true_divide

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-33-ccd86024eefa> in <module>()
      2 
      3 # This first plot sets the frame
----> 4 frame.plot(florence_hyper, label="Florence")
      5 
      6 # And then we add the Moon

~/Development/poliastro/poliastro-library/src/poliastro/plotting.py in plot(self, orbit, label, color)
    169         self.set_attractor(orbit.attractor)
    170         self._redraw_attractor(orbit.r_p * 0.15)  # Arbitrary Threshhold
--> 171         _, positions = orbit.sample(self.num_points)
    172 
    173         x0, y0 = self._project(orbit.r[None])

~/Development/poliastro/poliastro-library/src/poliastro/twobody/orbit.py in sample(self, values, method)
    331                 nu_values = np.linspace(-nu_limit, nu_limit, values)
    332 
--> 333             return self.sample(nu_values, method)
    334 
    335         elif hasattr(values, "unit") and values.unit in ('rad', 'deg'):

~/Development/poliastro/poliastro-library/src/poliastro/twobody/orbit.py in sample(self, values, method)
    336             values = self._generate_time_values(values)
    337 
--> 338         return values, self._sample(values, method)
    339 
    340     def _sample(self, time_values, method=mean_motion):

~/Development/poliastro/poliastro-library/src/poliastro/twobody/orbit.py in _sample(self, time_values, method)
    341         values = np.zeros((len(time_values), 3)) * self.r.unit
    342         for ii, epoch in enumerate(time_values):
--> 343             rr = self.propagate(epoch, method).r
    344             values[ii] = rr
    345 

~/Development/poliastro/poliastro-library/src/poliastro/twobody/orbit.py in propagate(self, value, method, rtol, **kwargs)
    273                 time_of_flight = time.TimeDelta(value)
    274 
--> 275             return propagate(self, time_of_flight, method=method, rtol=rtol, **kwargs)
    276 
    277     def sample(self, values=None, method=mean_motion):

~/Development/poliastro/poliastro-library/src/poliastro/twobody/propagation.py in propagate(orbit, time_of_flight, method, rtol, **kwargs)
    207 
    208     """
--> 209     r, v = method(orbit, time_of_flight.to(u.s).value, rtol=rtol, **kwargs)
    210     return orbit.from_vectors(orbit.attractor, r * u.km, v * u.km / u.s, orbit.epoch + time_of_flight)
    211 

~/Development/poliastro/poliastro-library/src/poliastro/twobody/propagation.py in mean_motion(orbit, tof, **kwargs)
    140         with u.set_enabled_equivalencies(u.dimensionless_angles()):
    141             M = M0 + tof * np.sqrt(k / np.abs(a ** 3)) * u.rad
--> 142             nu = M_to_nu(M, ecc)
    143 
    144     # parabolic orbit

~/Development/poliastro/poliastro-library/src/poliastro/twobody/angles.py in M_to_nu(M, ecc)
    231     """
    232     if ecc > 1:
--> 233         F = M_to_F(M, ecc)
    234         nu = F_to_nu(F, ecc)
    235     else:

~/Development/poliastro/poliastro-library/src/poliastro/twobody/angles.py in M_to_F(M, ecc)
    158     with u.set_enabled_equivalencies(u.dimensionless_angles()):
    159         F = optimize.newton(_kepler_equation_hyper, M, _kepler_equation_prime_hyper,
--> 160                             args=(M, ecc))
    161     return F
    162 

~/.miniconda36/envs/poliastro36/lib/python3.6/site-packages/scipy/optimize/zeros.py in newton(func, x0, fprime, args, tol, maxiter, fprime2)
    213             q1 = func(*((p1,) + args))
    214     msg = "Failed to converge after %d iterations, value is %s" % (maxiter, p)
--> 215     raise RuntimeError(msg)
    216 
    217 

RuntimeError: Failed to converge after 50 iterations, value is nan rad

```

Doing some debugging, it seems that it's trying to sample in values that are too far away in time. For example, for `florence_hyper.sample(1)`:

```
> /home/juanlu/Development/poliastro/poliastro-library/src/poliastro/twobody/orbit.py(343)_sample()
    341         values = np.zeros((len(time_values), 3)) * self.r.unit
    342         for ii, epoch in enumerate(time_values):
--> 343             rr = self.propagate(epoch, method).r
    344             values[ii] = rr
    345 

ipdb> p time_values
<Time object: scale='tdb' format='iso' value=['1856-07-27 09:14']>
```

A workaround would be changing the sampling method to `kepler`, which works without problems in this particular case, but the current plotting interface does not allow it (this is a separate issue).

📋  **Steps to solve the problem**

 * Comment below about what you've started working on.
 * Add, commit, push your changes
 * Submit a pull request and add this in comments - `Addresses #<put issue number here>`
 * Ask for a review in comments section of pull request
 * Celebrate your contribution to this project 🎉

**Comment by shreyasbapat** [comment]
So, basically, changing the default propagator from ``mean_motion`` to ``kepler`` is the workaround?

**Comment by Juanlu001** [comment]
> So, basically, changing the default propagator from mean_motion to kepler is the workaround?

No, because then we would be back to the problems `kepler` had with some other orbits.

> What you basically suggest is that we must be able to pick up a certain propagator from OrbitPlotter itself?

Yes!

**Comment by Juanlu001** [comment]
But this is just a temporary workaround while we figure out what happens with `mean_motion`.

**Comment by shreyasbapat** [comment]
Fine! So, passing the propagator as an argument is fine? in ``OrbitPlotter.plot()`` maybe?

**Comment by Juanlu001** [comment]
I'd rather add it in the `OrbitPlotter` constructor, but let me comment on your PR. Thanks!

**Comment by shreyasbapat** [comment]
Anywhere... I thought by adding it to plot, we can control different propagators for different orbits

**Comment by nikita-astronaut** [comment]
Hello everyone! @shreyasbapat , in your pull request, are you going to pass the `kepler` propagation method to `.plot` in Florence notebook? Or I will do it in the other pull request after you add the option to pass propagation method to Orbit?

Can I actually start before you merge? Or this won't do on github?

**Comment by Juanlu001** [comment]
I clarified offline with @nikita-astronaut that we have to treat the
convergence failure separately, so don't worry about this PR. I am not sure
myself about the approach here, would like to compare with GMAT or STK to
see how do they set up propagators for a simulation (per-simulation or
per-orbit).

**Comment by shreyasbapat** [comment]
So we leave it for now (atleast)

**Comment by Juanlu001** [comment]
I just remembered that, in STK, you can indeed select a different propagator for each orbit in one scenario:

![frame](https://user-images.githubusercontent.com/316517/39110519-b6c7c4c4-46d1-11e8-91df-bf96b9c34626.png)

So I've changed my mind to accept this. @shreyasbapat would you like to change the propagator in the last orbit of the Florence notebook to fix it? (See the errors here http://docs.poliastro.space/en/latest/examples/Catch%20that%20asteroid!.html)

**Comment by Juanlu001** [comment]
This was fixed in #353, merged manually. Closing!
